x/concentrated-liquidity/lp.go

```diff
@@ -17,88 +17,152 @@ import (
+	if err != nil {
+		return sdk.Int{}, sdk.Int{}, err
+	}
+
+	pool.updateLiquidityIfActivePosition(ctx, lowerTick, upperTick, liquidityDelta)
+
+	k.setPoolById(ctx, pool.Id, pool)
+
+	return actualAmount0, actualAmount1, nil
+}
+
+// calcActualAmounts calculates and returns actual amounts based on where the current tick is located relative to position's
+// lower and upper ticks.
+// The returned amounts are rounded down to avoid returning more to clients than they actually deposited.
+// There are 3 possible cases:
+// -The position is active ( lowerTick <= p.CurrentTick < upperTick).
+//    * The provided liqudity is distributed in both tokens.
+//    * Actual amounts might differ from desired because we recalculate them from liquidity delta and sqrt price.
+//      the calculations lead to amounts being off. // TODO: confirm logic is correct
+// - Current tick is below the position ( p.CurrentTick < lowerTick).
+//    * The provided liquidity is distributed in token0 only.
+// - Current tick is above the position ( p.CurrentTick >= p.upperTick ).
+//    * The provided liquidity is distributed in token1 only.
+// TODO: add tests.
+func (p Pool) calcActualAmounts(ctx sdk.Context, lowerTick, upperTick int64, sqrtRatioLowerTick, sqrtRatioUpperTick sdk.Dec, liquidityDelta sdk.Dec) (actualAmountDenom0 sdk.Int, actualAmountDenom1 sdk.Int) {
+	if p.isCurrentTickInRange(lowerTick, upperTick) {
+		// outcome one: the current price falls within the position
 		// if this is the case, we attempt to provide liquidity evenly between asset0 and asset1
 		// we also update the pool liquidity since the virtual liquidity is modified by this position's creation
-		amtDenom0 = calcAmount0Delta(liquidity, currentSqrtPrice, sqrtRatioUpperTick, false).RoundInt()
-		amtDenom1 = calcAmount1Delta(liquidity, currentSqrtPrice, sqrtRatioLowerTick, false).RoundInt()
-		pool.Liquidity = pool.Liquidity.Add(liquidity)
+		currentSqrtPrice := p.CurrentSqrtPrice
+		actualAmountDenom0 = calcAmount0Delta(liquidityDelta, currentSqrtPrice, sqrtRatioUpperTick, false).TruncateInt()
+		actualAmountDenom1 = calcAmount1Delta(liquidityDelta, currentSqrtPrice, sqrtRatioLowerTick, false).TruncateInt()
+	} else if p.CurrentTick.LT(sdk.NewInt(lowerTick)) {
+		// outcome two: position is below current price
+		// this means position is solely made up of asset0
+		actualAmountDenom1 = sdk.ZeroInt()
+		actualAmountDenom0 = calcAmount0Delta(liquidityDelta, sqrtRatioLowerTick, sqrtRatioUpperTick, false).TruncateInt()
 	} else {
 		// outcome three: position is above current price
 		// this means position is solely made up of asset1
-		amtDenom0 = sdk.ZeroInt()
-		amtDenom1 = calcAmount1Delta(liquidity, sqrtRatioLowerTick, sqrtRatioUpperTick, false).RoundInt()
+		actualAmountDenom0 = sdk.ZeroInt()
+		actualAmountDenom1 = calcAmount1Delta(liquidityDelta, sqrtRatioLowerTick, sqrtRatioUpperTick, false).TruncateInt()
```

Comment:
The only thing I am wondering here is if this needs to be an Int? I guess my question is, is this an intermediate step? If so, shouldn't we lean towards Decs and then only truncate at the very last step? If this is the "last step", then this is fine.

Reply:
This is the last step before being returned to the user.

I agree though that truncating here is awkward, and I just copied the pre-existing logic.

I moved the truncation higher up the call stack to the `updatePosition` function